builder/build_validate.go

```diff
@@ -254,46 +263,74 @@ func (b *Builder) mcaPkgInfo(mInfo map[string]*swupd.Manifest, version, download
 	pInfo := make(map[string]*mcaBundlePkgInfo)
 
 	// Download and query file metadata from all packages in each bundle
+	bundleWorker := func() {
+		for m := range mCh {
+			var pkgList = []string{}
+			for pkg := range mInfo[m.name].BundleInfo.AllPackages {
+				pkgList = append(pkgList, pkg)
+			}
+
+			out, err := downloadRpms(packagerCmd, pkgList, buildVersionDir, downloadRetries)
+			if err != nil {
+				errorCh <- err
+				wg.Done()
+				return
+			}
+
+			// Collect metadata to resolve installed RPM file names
+			bundlePkgInfo := pkgInfoFromNoopInstall(out.String())
+			for _, p := range bundlePkgInfo {
+				mux.Lock()
```

Comment:
Taking another look at this lock block, noticed that if one of the goroutines hit a package that is not yet part of the cache, all the other goroutines will be locked, even the ones trying to access packages that are already cached.
To prevent that, we could change to a RWLock, which would allow us to run `resolvePkgFiles` outside the lock context and let the other goroutines proceed. The downside of this approach is that while a goroutine is not done processing a package, another one might start the processing the same package in parallel. In that case, when we acquire the write lock we need to reevaluate `if pkgCache[p.name] == nil` before updating `pkgCache`.
All of this is under the assumption that  `resolvePkgFiles` is a costly operation given that it relies on `exec()`. It might be nice to compare both to have an idea of the impact in the overall time.

Reply:
I'll investigate this.